types/project.go

```diff
@@ -186,21 +186,16 @@ func (p *Project) AllServices() Services {
 	return all
 }
 
-type ServiceFunc func(name string, service ServiceConfig) error
+type ServiceFunc func(name string, service *ServiceConfig) error
 
-// WithServices runs ServiceFunc on each service and dependencies according to DependencyPolicy
-// It returns a new Project instance with the changes and keep the original Project unchanged
-func (p *Project) WithServices(names []string, fn ServiceFunc, options ...DependencyOption) (*Project, error) {
-	newProject, err := p.deepCopy()
-	if err != nil {
-		return nil, err
-	}
+// ForEachService runs ServiceFunc on each service and dependencies according to DependencyPolicy
+// The fn ServiceFunc should not mutate its ServiceConfig attribute
```

Comment:
```suggestion
// ForEachService runs ServiceFunc on each service and dependencies according to DependencyPolicy.
//
// No copies are created; any mutation of `*ServiceConfig` by the fn `ServiceFunc` will apply to this `*Project` object.
```

My 2c: either we should pass `ServiceConfig::deepCopy()` to `ServiceFunc` to make it impossible for callers to mutate the project, or accept that this function allows for in-place mutation and encourage callers to call `proj.Clone()` themselves if needed.

Reply:
we can indeed clone the ServiceConfig as a foolguard, anyway any attempt we make in this direction can only be a recommendation, as attributes are exported so third party can do mutation in many places (so does docker/compose, we don't use this function so much, and have tons of `for s := range project.Services`)

Reply:
@milas done, I added a `ServiceConfig.deepCopy` function and use it when calling the `fn` ServiceFunc 👍